### ISSUE TITLE
Issues with slowly opening findbar.

### ISSUE THREAD

**Issue by mzso** [Quicksaver/FindBar-Tweak#198]
I don't know where the issue lies but in recent weeks (months) the findbar tends to open rather slowly. Which wouldn't be a problem, but I type before it's properly opened which results in abnormal things. Sometimes the first few characters I typed are lost. Other times the findbar is not cleared as I type and the characters end up preceding whatever the content of the findbar was before. There were a couple of times when the findbar became red for a second or two even though what I typed was on the page, but as I kept typing it actually found a result and got back to normal.

**Comment by Quicksaver** [comment]
Yeah you're right, I've experienced this too a few times. I've been trying to figure out exactly what is happening but I have to admit I haven't been able to fix it yet. Any keystrokes should only be registered after the findbar opens (even if you press the keys before the findbar is actually visible), because all these actions are performed sequentially.

It's a bit weird that this happens, but of course that it shouldn't, and I'm going to try and fix it as soon as I can.

**Comment by Quicksaver** [comment]
When it happens for me, it's usually soon after starting up firefox. It doesn't seem to happen after a while, even in new tabs. Can you tell if this is your case as well?

**Comment by mzso** [comment]
I didn't notice any pattern. But I might be experiencing more often. I have an aging PC and a heavy profile, so things like this might come up more often.

**Comment by mzso** [comment]
Isn't it possible to speed up the showing of the findbar? Judging by the often quite significant delay compared to me starting to type in a normal place, I'm assuming a bunch of code runs before the findbar is actually opened.

Alternatively couldn't keystrokes be buffered asynchronously before the findbar is actually shown?

**Comment by Quicksaver** [comment]
> Judging by the often quite significant delay compared to me starting to type in a normal place, I'm assuming a bunch of code runs before the findbar is actually opened.

Not exactly, actually opening the findbar and placing the cursor in it is (almost) very first thing done. The problem is that, because of e10s, the part of prefilling it with the selected text in the webpage is done in an asynchronous step (even when there is no selected text, this step is still done to "prefill it with emptiness" basically). This bug is caused when you open the findbar and type in it before it is prefilled.

I made a small change to make it so it doesn't prefill after you've typed, which while by itself shouldn't really speed things up, it should at least prevent the bug from happening letting you use the findbar as you normally would from the start.

I did make some changes to the global findbar so that it is opened (and closed) faster. I don't know if you're using it, but if you are you definitely should see some improvement there (hopefully).

Please give beta version v2.0.1b1 a try (https://github.com/Quicksaver/FindBar-Tweak/releases/tag/v2.0.1b1) and let me know how it works for you.

**Comment by mzso** [comment]
Ummm. I'm experiencing this issue without e10s enabled on ff 38/39.

I mentioned like three manifestations, which are the ones you're also experiencing?

I'll test v2.0.1b1.

PS: actually I have it prefilled with text even if I don't have anything selected. You said it was because of my custom hotkeys: https://github.com/Quicksaver/FindBar-Tweak/issues/187#issuecomment-111882498

**Comment by Quicksaver** [comment]
> Ummm. I'm experiencing this issue without e10s enabled on ff 38/39.

The mechanism is e10s-ready is what I meant, basically it will work the same way in e10s and non-e10s, except of course in non-e10s everything runs in the same thread, so it can overload the processor much more easily and lead to situations like these more often.

> I mentioned like three manifestations, which are the ones you're also experiencing?

All of them. :) If I'm right about what's happening, those are all symptoms of the same underlying problem: typing between opening the findbar and it finishing the whole opening process, it just produces a different symptom depending on exactly where in the sequence of events the typing occurs. Of course I could be wrong, so let me know if that beta helps at all with anything and I'll look again.

> PS: actually I have it prefilled with text even if I don't have anything selected. You said it was because of my custom hotkeys: #187 (comment)

Hmm I had forgotten about that. I'm not sure if that changes things though. My guess is, if there's still any issue to be fixed, it would be that the previous text is kept after the cursor as you type instead of being replaced, maybe... (hard to say as I didn't test that scenario) I guess in this case there's no one better than you to try it and test it out. :)

**Comment by mzso** [comment]
For now I can say that in this short time I experienced the first two for sure. I don't remember the false "turning red" but it might have happened also.

**Comment by Quicksaver** [comment]
I admit I did not expect that... Can you tell if it happens as often as before?

**Comment by mzso** [comment]
I can't tell. Doesn't feel different.

**Comment by Quicksaver** [comment]
Nothing's ever simple to fix... :( It seems this might take a little while to fix, as I need to figure out what could even be happening. I'm sorry about that, and I hope it's not a big bother in the meantime.

**Comment by mzso** [comment]
I see you committed relevant stuff. I'm going to test it.

**Comment by Quicksaver** [comment]
There's no need, that's what was in the beta I linked you to before, I just had forgotten to sync at the time and only noticed now.

**Comment by Quicksaver** [comment]
Could you see if beta 2.0.1b2 makes any difference? (https://github.com/Quicksaver/FindBar-Tweak/releases/tag/v2.0.1b2) I ran into a piece of code in the findbar initialization that could in theory be relevant, I doubt it's what is causing this, but who knows.

**Comment by mzso** [comment]
Sure.

(I moved to nightly/electrolysis experimentally for now, so it might not present itself as it does on stable/beta)

**Comment by Quicksaver** [comment]
Forget about v2.0.1b2, please try v2.0.1b3 (https://github.com/Quicksaver/FindBar-Tweak/releases/tag/v2.0.1b3).

I think I figured it out, when you opened the findbar and started typing too fast, you were actually typing in the content itself because the findbar wasn't open yet, so each of that keypress was re-opening/re-erasing/re-setting/re-focusing/re-something the findbar. I fixed it so it checks on each keypress if the findbar is already opened (I think in theory this also happens in Firefox without FBT, if you confirm that this beta fixes it, I'll have to open a bug to let them know). Let's see if this time I got it right.

By the way, also fixed the phrase not found initial status, turns out it was a completely different bug.

**Comment by mzso** [comment]
I was just about to report that I noticed the issue once under e10s.
So I guess I'll start testing this then.

> I fixed it so it checks on each keypress if the findbar is already opened

So this means those keypresses are still lost, right? (Unless you save/buffer them)

**Comment by Quicksaver** [comment]
No no, they just check if they should reopen/refocus/etc the findbar. Even if they don't need to, they are still carried over to it and processed accordingly (well, that's the intention at least).

**Comment by Quicksaver** [comment]
TL;DR: (in case you're interested in the long explanation)

For instance, say you type "/", "a", "b", "c".

In most cases, what happens is:
"/" is typed in the webpage - opens findbar, select its contents, place cursor in it - "a" is typed already in the findbar so it replaces its contents, then "b", then "c"

But if you're fast enough, which in your case seems to happen more easily, it may happen that all those keys are pressed in the webpage, basically "a", "b" and "c" are not typed in the findbar itself, but in the webpage because the findbar took too long to open.

Each of those keys is supposed to focus the findbar if isn't already focused, and because they were typed in the webpage, it just assumes they have to focus the findbar without checking if a previous key already did that. So you get a mixture of having the cursor in the beginning or middle of the find query, or only a subset of those keys appearing in the end query, because those initial keys kept resetting or refocusing the findbar when they shouldn't.

I just added a sort of "did a previous key focus the findbar?" check, so subsequent keypresses happen as if they were typed directly in the findbar. And of course, after the findbar is already opened, any other keypresses are added directly to it, no such checks are necessary.

Of course, this is a lot of assuming on my part, you're the one who holds the verdict, I actually haven't seen this issue myself (except the phrase not found wrong status) since before you opened the issue. :)

**Comment by mzso** [comment]
Thanks for the explanation.

**Comment by mzso** [comment]
No joy it seems. I experienced the new/old search term mixing a number of times. The falsely red thing maybe once or twice. Don't remember any typed characters missing though.

**Comment by Quicksaver** [comment]
Bugger...

**Comment by Quicksaver** [comment]
> Don't remember any typed characters missing though.

Better than nothing I guess... 

I don't think I've asked you this (if I have, sorry, please remind me), do you have "Fill the findbar with the selected text" enabled? (on "When selecting text", not on "keyboard shortcut")

**Comment by mzso** [comment]
In that section it's unticked. I don't think I've ever used that feature.

**Comment by mzso** [comment]
Couldn't it be (in some part at least) related to the feature of the findbar which stores, then restores a search term I used in a different tab.
It feels like to me it mostly happens when I open the findbar on a particular tab the first time.

**Comment by Quicksaver** [comment]
That's what I was thinking. I looked into that code before but nothing that could be causing this really stood out at the time, so I moved on to what I did in the last beta I linked you to. I definitely will look into that part again more carefully.

**Comment by Quicksaver** [comment]
Let's test out that theory. Please try v2.0.1b4 and let me know how it goes (https://github.com/Quicksaver/FindBar-Tweak/releases/tag/v2.0.1b4)

**Comment by mzso** [comment]
So far I saw the "false red" thing once and the old term not clearing thing 2-3 times.
The missing typed characters thing still didn't show. So I guess that's fixed now.

**Comment by Quicksaver** [comment]
Ok let's see the state of things.

- Missing first typed characters: fixed, yay!

- False "Phrase not found" state: there are dozens of methods and functions and actions that can trigger a status change. Since this isn't really critical because it changes to the correct status after a moment, unless you're able to give me exact steps to reproduce it, I don't think I'll be going through every possibility here. So at least for now let's focus on the last point.

- Old find query not clearing: I'm running out of ideas... One more try: https://github.com/Quicksaver/FindBar-Tweak/releases/tag/v2.0.1b5

**Comment by mzso** [comment]
Okay. I'm testing it.

**Comment by mzso** [comment]
I didn't browse too actively the past few days (lack of time), but I didn't see any of the issues pop up. So far so good.

**Comment by Quicksaver** [comment]
One can hope. :) Let me know if that changes.

**Comment by Quicksaver** [comment]
I'm uploading the next version later today with all the fixes in this issue, so I'm considering everything fixed as we discussed so far. Feel free to re-open or create a new issue if it starts acting up again.

**Comment by mzso** [comment]
I think it's fixed. None happens anymore.
